cmd/soroban-rpc/lib/preflight/src/lib.rs

```diff
@@ -0,0 +1,232 @@
+    pub base_reserve: u32,
+}
+
+impl From<CLedgerInfo> for LedgerInfo {
+    fn from(c: CLedgerInfo) -> Self {
+        let network_passphrase_cstr = unsafe { CStr::from_ptr(c.network_passphrase) };
+        Self {
+            protocol_version: c.protocol_version,
+            sequence_number: c.sequence_number,
+            timestamp: c.timestamp,
+            network_passphrase: network_passphrase_cstr
+                .to_str()
+                .unwrap()
+                .as_bytes()
+                .to_vec(),
+            base_reserve: c.base_reserve,
+        }
+    }
+}
+
+fn storage_footprint_to_ledger_footprint(
+    foot: &storage::Footprint,
+) -> Result<LedgerFootprint, xdr::Error> {
+    let mut read_only: Vec<LedgerKey> = Vec::new();
+    let mut read_write: Vec<LedgerKey> = Vec::new();
+    for (k, v) in &foot.0 {
+        match v {
+            AccessType::ReadOnly => read_only.push((**k).clone()),
+            AccessType::ReadWrite => read_write.push((**k).clone()),
+        }
+    }
+    Ok(LedgerFootprint {
+        read_only: read_only.try_into()?,
+        read_write: read_write.try_into()?,
+    })
+}
+
+#[repr(C)]
+pub struct CPreflightResult {
+    pub error: *mut libc::c_char, // Error string in case of error, otherwise null
+    pub result: *mut libc::c_char, // SCVal XDR in base64
+    pub footprint: *mut libc::c_char, // LedgerFootprint XDR in base64
+    pub cpu_instructions: u64,
+    pub memory_bytes: u64,
+}
+
+fn preflight_error(str: String) -> *mut CPreflightResult {
+    let c_str = CString::new(str).unwrap();
+    // transfer ownership to caller
+    // caller needs to invoke free_preflight_result(result) when done
+    Box::into_raw(Box::new(CPreflightResult {
+        error: c_str.into_raw(),
+        result: null_mut(),
+        footprint: null_mut(),
+        cpu_instructions: 0,
+        memory_bytes: 0,
+    }))
+}
+
+#[no_mangle]
+pub extern "C" fn preflight_host_function(
+    handle: libc::uintptr_t, // Go Handle to forward to SnapshotSourceGet and SnapshotSourceHasconst
+    hf: *const libc::c_char, // HostFunction XDR in base64
+    source_account: *const libc::c_char, // AccountId XDR in base64
+    ledger_info: CLedgerInfo,
+) -> *mut CPreflightResult {
+    // catch panics before they reach foreign callers (which otherwise would result in
+    // undefined behavior)
+    let res = panic::catch_unwind(panic::AssertUnwindSafe(|| {
+        preflight_host_function_or_maybe_panic(handle, hf, source_account, ledger_info)
+    }));
+    match res {
+        Err(panic) => match panic.downcast::<String>() {
+            Ok(panic_msg) => preflight_error(format!(
+                "panic during preflight_host_function() call: {}",
+                panic_msg
+            )),
+            Err(_) => preflight_error(
+                "panic during preflight_host_function() call: unknown cause".to_string(),
+            ),
+        },
+        Ok(r) => r,
+    }
+}
+
+fn preflight_host_function_or_maybe_panic(
```

Comment:
Intuitively, I feel like pushing the panic-handling down as low as possible, and using proper rust error handling would make this code easier to follow? But that might not be worth the effort...

Reply:
I think I've improved it, PTAL